In [1]:
#!/usr/bin/env python

# Imports
import numpy as np
import time

from sklearn import linear_model

### Nesta tarefa será implementada o algoritmo de regressão linear com múltiplas variáveis, utilizando a versão vetorizada.

### Os dados treinados serão de notas de alunos da UFCG, onde queremos prever o CRA, baseado em algumas disciplinas do primeiro período.

In [2]:
def compute_mse_vectorized(w, H, Y):
    res = Y - np.dot(H, w)
    totalError = np.dot(res.T, res)
    return (totalError / float(len(Y)))[0][0]

In [3]:
def step_gradient_vectorized(w_current, H, Y, learningRate):
    # y - Hw
    res = Y - np.dot(H, w_current)
    
    # −2HT (y−Hw)
    gradiente = np.multiply(-2, np.dot(H.T, res))
    
    # alpha * (−2HT (y−Hw))
    gradiente_atenuado = np.multiply(learningRate, gradiente)
    
    # w(t+1) = w(t) - alpha * (−2HT (y−Hw))
    new_w = np.subtract(w_current, gradiente_atenuado)
    
    return [new_w, gradiente]

In [4]:
def gradient_descent_runner_vectorized(starting_w, X,Y, learning_rate, epsilon):
    w = starting_w
    grad = np.array([np.inf] * len(X[0]))
    i = 0
    while (np.linalg.norm(grad) >= epsilon):
        w, grad = step_gradient_vectorized(w, X, Y, learning_rate)
        
        if i % 50000 == 0:
            print("MSE na iteração {0} é de {1}".format(i,compute_mse_vectorized(w, X, Y)))
        i+= 1
    return w

In [5]:
points = np.genfromtxt("data/sample_treino.csv", delimiter=",")
points = np.c_[np.ones(len(points)),points][1::]
X = points[:, :-1]
Y = points[:, -1:]
init_w = np.zeros((len(points[0]) - 1,1))
learning_rate = 0.00001
#num_iterations = 10000
epsilon = 0.05
print("Starting gradient descent with error = {0}".format(compute_mse_vectorized(init_w, X,Y)))
print("Running...")
tic = time.time()
w = gradient_descent_runner_vectorized(init_w, X,Y, learning_rate, epsilon)
toc = time.time()
print("Gradiente descendente convergiu com w0 = {0}, w1 = {1}, w2 = {2}, w3 = {3}, w4 = {4}, w5 = {5}, error = {6}".format(w[0], w[1], w[2], w[3], w[4], w[5], compute_mse_vectorized(w,X,Y)))
print("Versão vetorizada rodou em: " + str(1000*(toc-tic)) + " ms")

Starting gradient descent with error = 54.4799538556
Running...
MSE na iteração 0 é de 13.5401136888
MSE na iteração 50000 é de 0.417316489568
MSE na iteração 100000 é de 0.413093359157
MSE na iteração 150000 é de 0.411853190403
MSE na iteração 200000 é de 0.411489001191
MSE na iteração 250000 é de 0.411382053018
MSE na iteração 300000 é de 0.411350646509
Gradiente descendente convergiu com w0 = [1.69701142], w1 = [0.10377072], w2 = [0.04829928], w3 = [0.16400561], w4 = [0.38324514], w5 = [0.02077901], error = 0.411349181254
Versão vetorizada rodou em: 7909.50584412 ms


### Pode ser visto que a fórmula da regressão linear ficou:
CRA = 1.69701142 + 0.10377072 * Cálculo1 + 0.04829928 * LPT + 0.16400561 * P1 + 0.38324514 * IC + 0.02077901 * Cálculo2

### Agora serão comparados os coeficientes encontrados com o do scikit learn:

In [6]:
# Cria objeto de regressao linear
regr = linear_model.LinearRegression()

# Treina o modelo utilizando os dados de treino
regr.fit(X[:, 1:], Y)

print('Coefficients: ', regr.coef_)
print('Intercept: ', regr.intercept_)

('Coefficients: ', array([[0.10304143, 0.0464367 , 0.16409834, 0.38117843, 0.02027816]]))
('Intercept: ', array([1.73771151]))


### Pode ser visto que a fórmula da regressão linear ficou:
CRA = 1.73771151 + 0.10304143 * Cálculo1 + 0.0464367 * LPT + 0.16409834 * P1 + 0.38117843 * IC + 0.02027816 * Cálculo2

Apesar de não ter obtido os mesmos valores na implementação comparado com o scikit learn, os valores foram bastante similares, o que indica que o algoritmo foi implementado corretamente.